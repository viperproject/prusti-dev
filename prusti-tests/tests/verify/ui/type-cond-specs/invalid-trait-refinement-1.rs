use prusti_contracts::*;

trait HasContract {
    #[pure]
    fn pre(&self) -> bool;
    #[pure]
    fn post(&self) -> bool;
}

trait MyTrait {
    fn foo(&mut self);
}

#[extern_spec]
trait MyTrait {
    #[refine_spec(where Self: HasContract, [
        requires(self.pre()), ensures(self.post())
    ])]
    fn foo(&mut self);
}

struct MyStruct {
    x: i32,
}

#[refine_trait_spec]
impl MyTrait for MyStruct {
    #[requires(self.x >= 15)]
    fn foo(&mut self) {
        self.x += 10;
    }
}

impl HasContract for MyStruct {
    #[pure]
    fn post(&self) -> bool {
        self.x >= 20
    }

    #[pure]
    fn pre(&self) -> bool {
        self.x >= 10
    }
}

fn main() {
}
