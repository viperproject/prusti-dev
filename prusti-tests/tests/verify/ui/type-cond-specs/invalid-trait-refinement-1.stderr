error: [Prusti: verification error] the method's precondition may not be a valid weakening of the trait's precondition.
  --> $DIR/invalid-trait-refinement-1.rs:16:31
   |
16 |       #[refine_spec(where Self: HasContract, [
   |  _______________________________^
17 | |         requires(self.pre()), ensures(self.post())
   | |___________________________^
...
28 |       #[requires(self.x >= 15)]
   |                  ^^^^^^^^^^^^
   |
   = help: The trait's precondition should imply the implemented method's precondition.

error: aborting due to previous error

